### PR TITLE
CATO-4103 related party validation complete

### PR DESCRIPTION
**Pull request by pncampbell**

### CODE REVIEW

File: src/test/scala/uk/gov/hmrc/ct/accounts/frs10x/AccountsFreeTextValidationFixture.scala [hunk 1]
```diff
@@ -38,8 +38,13 @@ trait AccountsFreeTextValidationFixture extends WordSpec with Matchers with Mock
...
     }
 
+    testMandatoryAccountsCharacterLimitValidation(boxId, charLimit ,builder)
+  }
+
+  def testMandatoryAccountsCharacterLimitValidation(boxId: String, charLimit: Int, builder: (Option[String]) => ValidatableBox[AbridgedAccountsBoxRetriever]): Unit = {
+
     "pass validation when empty string" in {
-      builder(Some("")).validate(boxRetriever) shouldBe Set.empty
+      builder(Some("")).validate(boxRetriever) shouldBe Set()
```
Comment (liquidarmour) [inline src/test/scala/uk/gov/hmrc/ct/accounts/frs10x/AccountsFreeTextValidationFixture.scala]: .empty is better. No object creation.

File: src/main/scala/uk/gov/hmrc/ct/accounts/frs10x/abridged/relatedPartyTransactions/AC7803.scala [hunk 2]
```diff
@@ -30,8 +30,8 @@ case class AC7803(value: Option[String]) extends CtBoxIdentifier(name = "Name of
 
     collectErrors(
       validateAsMandatory(this),
-      validateOptionalStringByLength("AC7803", this, 1, StandardCohoTextFieldLimit),
-      validateOptionalStringByRegex("AC7803", this, ValidCoHoCharacters)
+      validateStringMaxLength("AC7803", value.getOrElse(""), StandardCohoTextFieldLimit),
```
Comment (liquidarmour) [inline src/main/scala/uk/gov/hmrc/ct/accounts/frs10x/abridged/relatedPartyTransactions/AC7803.scala]:
```
value.getOrElse("")
```
isn't as nice as the previous version. 

Can't we have a function that takes an option the ignores if empty?

File: src/test/scala/uk/gov/hmrc/ct/accounts/frs10x/abridged/relatedPartyTransactions/AC7802Spec.scala [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+class AC7802Spec extends WordSpec with MockitoSugar with Matchers with BeforeAndAfter
+  with MockAbridgedAccountsRetriever with AccountsFreeTextValidationFixture {
+
+  testMandatoryAccountsCharacterLimitValidation("AC7802", StandardCohoTextFieldLimit, AC7802)
+  testAccountsCoHoTextFieldValidation("AC7802", AC7802)
+
+  "AC7802" should {
+    "be mandatory" in {
+      AC7802(None).validate(boxRetriever) shouldBe Set(CtValidation(Some("AC7802"),"error.AC7802.required",None))
```
Comment (liquidarmour) [inline src/test/scala/uk/gov/hmrc/ct/accounts/frs10x/abridged/relatedPartyTransactions/AC7802Spec.scala]: spaces after commas

What about the valid case? Is that handled in the test functions above?

File: src/test/scala/uk/gov/hmrc/ct/accounts/frs10x/abridged/relatedPartyTransactions/AC7803Spec.scala [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+class AC7803Spec extends WordSpec with MockitoSugar with Matchers with BeforeAndAfter
+  with MockAbridgedAccountsRetriever with AccountsFreeTextValidationFixture {
+
+  testMandatoryAccountsCharacterLimitValidation("AC7803", StandardCohoTextFieldLimit, AC7803)
+  testAccountsCoHoTextFieldValidation("AC7803", AC7803)
+
+  "AC7803" should {
+    "be mandatory" in {
+      AC7803(None).validate(boxRetriever) shouldBe Set(CtValidation(Some("AC7803"),"error.AC7803.required",None))
```
Comment (liquidarmour) [inline src/test/scala/uk/gov/hmrc/ct/accounts/frs10x/abridged/relatedPartyTransactions/AC7803Spec.scala]: spaces

File: src/test/scala/uk/gov/hmrc/ct/accounts/frs10x/abridged/relatedPartyTransactions/AC7802Spec.scala (hunk 3 above)
Comment (pncampbell) [inline src/test/scala/uk/gov/hmrc/ct/accounts/frs10x/abridged/relatedPartyTransactions/AC7802Spec.scala]: The test functions test pass and fail scenarious

File: src/main/scala/uk/gov/hmrc/ct/accounts/frs10x/abridged/relatedPartyTransactions/AC7803.scala (hunk 2 above)
Comment (pncampbell) [inline src/main/scala/uk/gov/hmrc/ct/accounts/frs10x/abridged/relatedPartyTransactions/AC7803.scala]: I tried to change it as you suggested but it breaks the text validation fixtures in a way I cant (yet) work out how to fix. About 15 existing boxes use the same pattern (which is why I used it ....), validateOptionalStringByRegex doesnt work with the text validation fixture which is why I aligned with whats already done

**Pull request merged**